### PR TITLE
Dev deepfm

### PR DESCRIPTION
**Pull request by Liuxinman**
TODO:
- [ ] README.md
- [ ] Add clip gradient norm
- [ ] Add L2 regularization for OneEmbedding
- [ ] Add dropout in fused mlp

### CODE REVIEW

File: RecommenderSystems/deepfm/deepfm_train_eval.py [hunk 1]
```diff
@@ -0,0 +1,662 @@
...
+import sys
+import glob
+import time
+import math
+import numpy as np
+import psutil
+import warnings
+import oneflow as flow
+import oneflow.nn as nn
+from sklearn.metrics import roc_auc_score, log_loss
```
Comment (ShawnXuan) [inline RecommenderSystems/deepfm/deepfm_train_eval.py:12]: use flow.roc_auc_score
can log_loss be replaced with "flow.nn.BCELoss" ?

File: RecommenderSystems/deepfm/deepfm_train_eval.py [hunk 2]
```diff
@@ -0,0 +1,662 @@
...
+    train_graph = DeepFMTrainGraph(deepfm_module, loss, opt, grad_scaler, args.amp)
+
+    train_losses = []
+    batches_per_epoch = math.ceil(args.num_train_samples / args.batch_size)
+
+    best_metric = -np.inf
+    stopping_steps = 0
+
+    deepfm_module.train()
+    step, last_step, last_time = -1, 0, time.time()
```
Comment (ShawnXuan) [inline RecommenderSystems/deepfm/deepfm_train_eval.py:549]: move to before line 552

File: RecommenderSystems/deepfm/deepfm_train_eval.py [hunk 3]
```diff
@@ -0,0 +1,662 @@
...
+    deepfm_module.load_state_dict(state_dict, strict=False)
+    print("================ Validation Evaluation ================")
+    auc, logloss = eval(args, eval_graph, tag='val', cur_step=step, epoch=epoch)
+    print("================ Test Evaluation ================")
+    auc, logloss = eval(args, eval_graph, tag='test', cur_step=step, epoch=epoch)
+
+
+def batch_to_global(np_label, np_features):
+    def _np_to_global(np, dtype=flow.float):
+        t = flow.tensor(np, dtype=dtype)
```
Comment (ShawnXuan) [inline RecommenderSystems/deepfm/deepfm_train_eval.py:602]: use flow.from_numpy can reduce `cast` time, but the dtype in dataset should match with target dtype.

File: RecommenderSystems/deepfm/deepfm_train_eval.py (hunk 2 above)
Comment (Liuxinman) [reply RecommenderSystems/deepfm/deepfm_train_eval.py]: 已修改

File: RecommenderSystems/deepfm/deepfm_train_eval.py (hunk 1 above)
Comment (Liuxinman) [reply RecommenderSystems/deepfm/deepfm_train_eval.py]: 已改为用flow.roc_auc_score
sklearn的log_loss好像和nn.BCELoss并不等价，主要是在sklearn里probabilities are clipped to max(eps, min(1 - eps, p))。我之后会再仔细看看。
